src/OrleansRuntime/Streams/QueueBalancer/LeaseBasedQueueBalancer.cs

```diff
@@ -0,0 +1,345 @@
+            // Even if they are not yet, they will be mature by the time I mature myself (after I become !isStarting).
+            foreach (var silo in siloStatusOracle.GetApproximateSiloStatuses(true).Keys.Where(s => !s.Equals(siloStatusOracle.SiloAddress)))
+            {
+                immatureSilos[silo] = false;     // record as mature
+            }
+        }
+
+        public override Task Initialize(string strProviderName, IStreamQueueMapper queueMapper, TimeSpan siloMaturityPeriod)
+        {
+            if (queueMapper == null)
+            {
+                throw new ArgumentNullException("queueMapper");
+            }
+            this.allQueues = new ReadOnlyCollection<QueueId>(queueMapper.GetAllQueues().ToList());
+            this.siloMaturityPeriod = siloMaturityPeriod;
+            NotifyAfterStart().Ignore();
+            //make lease renew frequency to be every half of lease time, to avoid renew failing due to timing issues, race condition or clock difference. 
+            this.renewLeaseTimer = new AsyncTaskSafeTimer(this.MaintainAndBalanceQueues, null, this.siloMaturityPeriod, this.leaseLength.Divide(2));
+            //try to acquire maximum leases every leaseLength 
+            this.tryAcquireMaximumLeaseTimer = new AsyncTaskSafeTimer(this.AcquireLeaseToMeetMaxResponsibilty, null, this.siloMaturityPeriod, this.leaseLength);
+            //Selector default to round robin selector now, but we can make a further change to make selector configurable if needed.  Selector algorithm could 
+            //be affecting queue balancing stablization time in cluster initializing and auto-scaling
+            this.queueSelector = new RoundRobinSelector<QueueId>(this.allQueues.ToList());
+            return MaintainAndBalanceQueues(null);
+        }
+
+        public override IEnumerable<QueueId> GetMyQueues()
+        {
+            return this.myQueues.Select(queue => queue.QueueId);
+        }
+
+        private async Task MaintainAndBalanceQueues(object state)
+        {
+            CalculateResponsibility();
+            // step 1: renew existing leases 
+            await this.RenewLeases();
+            // step 2: if after renewing leases, myQueues count doesn't fall in [minimumResponsibility, maximumResponsibilty] range, act accordingly
+            if (this.myQueues.Count < this.minimumResponsibilty)
+            {
+                await this.AcquireLeasesToMeetMinResponsibility();
+            }
+            else if (this.myQueues.Count > this.maximumRespobsibility)
+            {
+                await this.ReleaseLeasesToMeetResponsibility();
+            }
+        }
+
+        private async Task ReleaseLeasesToMeetResponsibility()
+        {
+            var queueCountToRelease = this.myQueues.Count - this.maximumRespobsibility;
+            if (queueCountToRelease <= 0)
+                return;
+            var queuesToGiveUp = this.myQueues.GetRange(0, queueCountToRelease);
```

Comment:
I was suggesting using the Take(int) link function, as it would be more readable, but nm, this is more efficient anyway.